src/EditorFeatures/Core/Extensibility/SignatureHelp/SignatureHelpItem.cs

```diff
@@ -16,21 +17,21 @@ internal class SignatureHelpItem
         /// selected parameter index strictly goes past the number of defined parameters for this
         /// item.
         /// </summary>
-        public bool IsVariadic { get; private set; }
+        public bool IsVariadic { get; }
 
-        public IList<SymbolDisplayPart> PrefixDisplayParts { get; private set; }
-        public IList<SymbolDisplayPart> SuffixDisplayParts { get; private set; }
+        public ImmutableArray<SymbolDisplayPart> PrefixDisplayParts { get; }
+        public ImmutableArray<SymbolDisplayPart> SuffixDisplayParts { get; }
 
-        // TODO(cyrusn): This probably won't be sufficient for VB query signature help.  It has
+        // TODO: This probably won't be sufficient for VB query signature help.  It has
         // arbitrary separators between parameters.
-        public IList<SymbolDisplayPart> SeparatorDisplayParts { get; private set; }
+        public ImmutableArray<SymbolDisplayPart> SeparatorDisplayParts { get; }
 
-        public IList<SignatureHelpParameter> Parameters { get; private set; }
+        public ImmutableArray<SignatureHelpParameter> Parameters { get;  }
 
-        public IList<SymbolDisplayPart> DescriptionParts { get; internal set; }
+        public ImmutableArray<SymbolDisplayPart> DescriptionParts { get; internal set; }
 
-        // TODO(cyrusn): This may be unnecessary.  How would a user ever see this.
-        public IList<SymbolDisplayPart> Documentation { get; set; }
+        // Note: IEnumerable instead of ImmutableArray because we want lazy evaluation.
+        public IEnumerable<SymbolDisplayPart> Documentation { get; }
```

Comment:
That's somewhat concerning.  Couldn't this cause us to hold onto something like an IEnumerable and then evaluate it on the wrong threaD?  I recall issues like that in the past.

The way we dealt with this was to make it explicit that things were lazy by taking in ```Func<Whatever> createDocumentation```.  That way the caller understaood this would be called in the future.